File: android/biangbiang-ui/src/main/java/dev/veeso/biangbiangui/ui/screens/SettingsScreen.kt
package dev.veeso.biangbiangui.ui.screens

import android.content.Context
import android.content.Intent
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.DropdownMenu
import androidx.compose.material3.DropdownMenuItem
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SegmentedButton
import androidx.compose.material3.SegmentedButtonDefaults
import androidx.compose.material3.SingleChoiceSegmentedButtonRow
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.unit.dp
import androidx.core.net.toUri
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import dev.veeso.biangbiangui.config.SettingDescriptor
import dev.veeso.biangbiangui.services.ReviewPromptPolicy
import dev.veeso.biangbiangui.ui.AppDesign
import dev.veeso.biangbiangui.ui.BiangBiangContext
import dev.veeso.biangbiangui.ui.LocalBiangBiangContext
import java.util.Locale
import kotlinx.coroutines.launch

/**
 * The settings screen, restructured into the three spec layers. Config-driven
 * port of the reference BiangBiang Hanzi Android `SettingsModeView.kt`;
 * mirrors iOS Phase 3 `SettingsScreen`.
 *
 * 1. Shared core — translation-target picker (disabled as data when the active
 *    variant `translatable == false`), History Clear All, bug-report (GitHub
 *    issue URL from `branding.githubRepo`, email `branding.supportEmail`,
 *    subject `[Android] Bug report – <appName>`).
 * 2. Data-driven variant segmented picker — only when the active profile has
 *    more than one variant; binds `settings.selectedVariantId`. Changing it
 *    rebuilds the engine automatically via `BiangBiangContext` (the Android
 *    equivalent of iOS `rebuildEngine()` — the engine is derived from the
 *    selected-variant Flow).
 * 3. Injectable descriptors — one toggle/picker per `config.extraSettings`,
 *    bound to `settings.value`/`setValue`; footer from the descriptor.
 *
 * Keeps the DEBUG-only review-reset (developer-facing, never shipped).
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun SettingsScreen() {
    val ctx = LocalBiangBiangContext.current
    val context = LocalContext.current
    val scope = rememberCoroutineScope()

    val translatable = ctx.activeVariant?.translatable ?: true
    val history by ctx.settings.history.collectAsStateWithLifecycle(initialValue = emptyList())
    val userLanguage by ctx.settings.userLanguage
        .collectAsStateWithLifecycle(initialValue = "")
    val selectedVariantId by ctx.settings.selectedVariantId
        .collectAsStateWithLifecycle(
            initialValue = ctx.activeProfile.variants.firstOrNull()?.id ?: "",
        )

    var languageSelectExpanded by remember { mutableStateOf(false) }
    var showClearConfirm by remember { mutableStateOf(false) }

    val allLanguages = remember {
        Locale.getAvailableLocales()
            .filter { it.displayLanguage.isNotBlank() }
            .distinctBy { it.displayLanguage }
            .sortedBy { it.displayLanguage }
    }

    if (showClearConfirm) {
        AlertDialog(
            onDismissRequest = { showClearConfirm = false },
            title = { Text(ctx.config.strings.clearAllConfirm) },
            confirmButton = {
                TextButton(
                    onClick = {
                        scope.launch { ctx.settings.clearHistory() }
                        showClearConfirm = false
                    },
                    colors = ButtonDefaults.textButtonColors(
                        contentColor = MaterialTheme.colorScheme.error,
                    ),
                ) {
                    Text(ctx.config.strings.clearAll)
                }
            },
            dismissButton = {
                TextButton(onClick = { showClearConfirm = false }) {
                    Text("Cancel")
                }
            },
        )
    }

    Scaffold(
        topBar = { TopAppBar(title = { Text(ctx.config.strings.tabSettings) }) },
    ) { innerPadding ->
        Column(
            modifier = Modifier
                .padding(innerPadding)
                .verticalScroll(rememberScrollState())
                .padding(
                    horizontal = AppDesign.horizontalPadding,
                    vertical = AppDesign.sectionSpacing,
                )
                .fillMaxSize(),
            verticalArrangement = Arrangement.spacedBy(AppDesign.sectionSpacing),
        ) {
            // --- Layer 1: shared core ---
            SettingsSection(title = ctx.config.strings.translationLanguage) {
                Box(modifier = Modifier.fillMaxWidth()) {
                    OutlinedButton(
                        onClick = { languageSelectExpanded = true },
                        enabled = translatable,
                        modifier = Modifier.fillMaxWidth(),
                    ) {
                        Text(currentLanguageName(allLanguages, userLanguage))
                    }

                    DropdownMenu(
                        expanded = languageSelectExpanded,
                        onDismissRequest = { languageSelectExpanded = false },
                        modifier = Modifier.fillMaxWidth(0.9f),
                    ) {
                        allLanguages.forEach { locale ->
                            DropdownMenuItem(
                                text = { Text(locale.displayLanguage) },
                                onClick = {
                                    scope.launch {
                                        ctx.settings.setUserLanguage(locale.language)
                                    }
                                    languageSelectExpanded = false
                                },
                            )
                        }
                    }
                }
                if (!translatable) {
                    Text(
                        "Translation is unavailable for the selected variant.",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                }
            }

            // --- Layer 2: data-driven variant picker ---
            val variants = ctx.activeProfile.variants
            if (variants.size > 1) {
                SettingsSection(title = ctx.activeProfile.displayName) {
                    SingleChoiceSegmentedButtonRow(modifier = Modifier.fillMaxWidth()) {
                        variants.forEachIndexed { index, variant ->
                            SegmentedButton(
                                selected = selectedVariantId == variant.id,
                                onClick = {
                                    scope.launch {
                                        ctx.settings.setSelectedVariantId(variant.id)
                                    }
                                },
                                shape = SegmentedButtonDefaults.itemShape(
                                    index = index,
                                    count = variants.size,
                                ),
                                label = { Text(variant.displayName) },
                            )
                        }
                    }
                }
            }

            // --- Layer 3: injectable descriptors ---
            ctx.config.extraSettings.forEach { descriptor ->
                SettingsSection(title = descriptor.label) {
                    DescriptorControl(ctx, descriptor)
                    descriptor.footer?.let { footer ->
                        Text(
                            footer,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                }
            }

            // --- Library-owned History + support, after app-defined settings ---
            SettingsSection(title = ctx.config.strings.tabHistory) {
                OutlinedButton(
                    onClick = { showClearConfirm = true },
                    enabled = history.isNotEmpty(),
                ) {
                    Text(ctx.config.strings.clearAll)
                }
            }

            SettingsSection(title = ctx.config.strings.reportBug) {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalAlignment = Alignment.CenterVertically,
                ) {
                    Button(onClick = {
                        openGitHubIssues(context, ctx.config.branding.githubRepo)
                    }) {
                        Text(ctx.config.strings.openGithubIssues)
                    }
                    Button(onClick = {
                        sendBugEmail(
                            context,
                            ctx.config.branding.supportEmail,
                            ctx.config.branding.appName,
                        )
                    }) {
                        Text(ctx.config.strings.sendEmail)
                    }
                }
            }

            // --- DEBUG-only review reset ---
            if (isDebugBuild(context)) {
                val reviewLaunchCount by ctx.settings.reviewLaunchCount
                    .collectAsStateWithLifecycle(initialValue = 0)
                val reviewDismissed by ctx.settings.reviewPromptDismissed
                    .collectAsStateWithLifecycle(initialValue = false)
                val reviewAttempts by ctx.settings.reviewAttempts
                    .collectAsStateWithLifecycle(initialValue = 0)
                SettingsSection(title = "Debug") {
                    Text(
                        "review launchCount=$reviewLaunchCount dismissed=$reviewDismissed " +
                            "attempts=$reviewAttempts/${ReviewPromptPolicy.MAX_ATTEMPTS} " +
                            "(fires at ${ReviewPromptPolicy.LAUNCH_THRESHOLD})",
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.onSurfaceVariant,
                    )
                    Button(onClick = { scope.launch { ctx.settings.resetReviewPrompt() } }) {
                        Text("Reset review prompt")
                    }
                }
            }
        }
    }
}

@Composable
private fun DescriptorControl(
    ctx: BiangBiangContext,
    descriptor: SettingDescriptor,
) {
    val scope = rememberCoroutineScope()
    val current by produceState(initialValue = descriptor.defaultValue, descriptor.key) {
        value = ctx.settings.value(descriptor.key) ?: descriptor.defaultValue
    }
    when (val kind = descriptor.kind) {
        is SettingDescriptor.Kind.Toggle -> {
            Row(
                modifier = Modifier.fillMaxWidth(),
                horizontalArrangement = Arrangement.SpaceBetween,
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(descriptor.label)
                Switch(
                    checked = current == "true",
                    onCheckedChange = { checked ->
                        scope.launch {
                            ctx.settings.setValue(
                                descriptor.key,
                                if (checked) "true" else "false",
                            )
                        }
                    },
                )
            }
        }

        is SettingDescriptor.Kind.Picker -> {
            SingleChoiceSegmentedButtonRow(modifier = Modifier.fillMaxWidth()) {
                kind.options.forEachIndexed { index, option ->
                    SegmentedButton(
                        selected = current == option,
                        onClick = {
                            scope.launch { ctx.settings.setValue(descriptor.key, option) }
                        },
                        shape = SegmentedButtonDefaults.itemShape(
                            index = index,
                            count = kind.options.size,
                        ),
                        label = { Text(option) },
                    )
                }
            }
        }
    }
}

@Composable
private fun SettingsSection(title: String, content: @Composable () -> Unit) {
    Column(
        modifier = Modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(AppDesign.stackSpacing),
    ) {
        Text(title, style = MaterialTheme.typography.titleMedium)
        content()
    }
}

private fun currentLanguageName(locales: List<Locale>, current: String): String =
    locales.find { it.language == current }?.displayLanguage ?: current

private fun isDebugBuild(context: Context): Boolean =
    (context.applicationInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) != 0

private fun openGitHubIssues(context: Context, repo: String) {
    val intent = Intent(
        Intent.ACTION_VIEW,
        "https://github.com/$repo/issues/new".toUri(),
    )
    context.startActivity(intent)
}

private fun sendBugEmail(context: Context, email: String, appName: String) {
    val subject = "[Android] Bug report – $appName"
    val body = """
Description:

Steps to reproduce:

Device:
Android version:
"""
    val intent = Intent(Intent.ACTION_SENDTO).apply {
        data = "mailto:$email?subject=$subject&body=$body".toUri()
    }
    context.startActivity(Intent.createChooser(intent, "Send bug report"))
}

File: Sources/BiangBiangUI/UI/SettingsScreen.swift
//
//  SettingsScreen.swift
//  BiangBiangUI
//
//  Config-driven port of SettingsView from BiangBiang Hanzi, restructured
//  into the three spec layers:
//    1. Shared core — translation-target language picker (disabled as data
//       when the active variant `translatable == false`), History "Clear all",
//       and bug-report (GitHub issue URL + support email).
//    2. Data-driven variant segmented picker — rendered only when the active
//       profile has more than one variant; binds `settings.selectedVariantId`
//       and rebuilds the engine on change (the documented `rebuildEngine()`
//       contract — see BiangBiangContext).
//    3. Injectable descriptors — one toggle/picker per `config.extraSettings`,
//       bound to `settings.value(for:)` / `setValue(_:for:)`; footer from the
//       descriptor itself.
//  Generalisation notes vs. the reference:
//    • No hard-coded Chinese/Cantonese checks: the disabled-translation state
//      is driven by `activeVariant.translatable`; the variant list is data.
//    • User-facing strings come from ctx.config.strings everywhere except two
//      literals inherited verbatim from the reference + sibling precedent: the
//      confirmation-dialog "Cancel" button (identical to HistoryScreen) and the
//      translation-disabled footer (the reference SettingsView hard-coded it).
//      The GitHub repo and support email come from ctx.config.branding. The
//      DEBUG-only review-reset block is kept verbatim from the reference
//      (debug-only, developer-facing text — never shipped to end users).
//

#if canImport(UIKit)
    import SwiftUI

    // MARK: - SettingsScreen

    /// The settings screen: shared core controls, an optional data-driven variant
    /// picker, and injectable app-specific descriptors.
    ///
    /// Reads all configuration from `BiangBiangContext` in the SwiftUI environment.
    /// Inject it from the consuming app's root view:
    ///
    ///     ContentView()
    ///         .environment(context)
    @MainActor
    public struct SettingsScreen: View {
        @Environment(BiangBiangContext.self) private var ctx
        @Environment(\.openURL) private var openURL

        // MARK: - Private state

        @State private var showClearConfirm = false

        #if DEBUG
            @State private var showResetToast = false
        #endif

        // MARK: - Derived data

        /// Locale identifiers offered as translation targets, name-sorted.
        /// Mirrors the reference's `availableLanguages`; the selection binds the
        /// same `userLanguage` source the ported TextScreen translation reads.
        private static let availableLanguages: [(id: String, name: String)] =
            Locale.availableIdentifiers
                .map { id in
                    (
                        id: id,
                        name: Locale.current.localizedString(forIdentifier: id) ?? id
                    )
                }
                .sorted { $0.name < $1.name }

        // MARK: - Init

        public init() {}

        // MARK: - Body

        public var body: some View {
            @Bindable var settings = ctx.settings

            NavigationStack {
                Form {
                    sharedCoreSections(userLanguage: $settings.userLanguage)
                    variantSection(selectedVariantId: $settings.selectedVariantId)
                    descriptorSections()

                    #if DEBUG
                        debugSection
                    #endif
                }
                .navigationTitle(ctx.config.strings.tabSettings)
                .navigationBarTitleDisplayMode(.inline)
                #if DEBUG
                    .overlay(alignment: .bottom) {
                        if showResetToast {
                            CopyToast(message: "Review prompt reset")
                                .padding(.bottom, 24)
                                .transition(.move(edge: .bottom).combined(with: .opacity))
                                .task {
                                    try? await Task.sleep(for: .seconds(2))
                                    showResetToast = false
                                }
                        }
                    }
                    .animation(.easeInOut, value: showResetToast)
                #endif
            }
        }

        // MARK: - Layer 1: shared core

        @ViewBuilder
        private func sharedCoreSections(userLanguage: Binding<String>) -> some View {
            let translatable = ctx.activeVariant?.translatable ?? true

            Section {
                Picker(
                    ctx.config.strings.translationLanguage,
                    selection: userLanguage
                ) {
                    ForEach(Self.availableLanguages, id: \.id) { option in
                        Text(option.name).tag(option.id)
                    }
                }
                .pickerStyle(.menu)
                .disabled(!translatable)
            } header: {
                Label(ctx.config.strings.translationLanguage, systemImage: "globe")
            } footer: {
                if !translatable {
                    Text("Translation is unavailable for the selected variant.")
                }
            }

            Section {
                Button(ctx.config.strings.clearAll, role: .destructive) {
                    showClearConfirm = true
                }
                .disabled(ctx.settings.history.isEmpty)
            } header: {
                Label(ctx.config.strings.tabHistory, systemImage: "clock")
            }
            .confirmationDialog(
                ctx.config.strings.clearAllConfirm,
                isPresented: $showClearConfirm,
                titleVisibility: .visible
            ) {
                Button(ctx.config.strings.clearAll, role: .destructive) {
                    ctx.settings.clearHistory()
                }
                Button("Cancel", role: .cancel) {}
            }

            Section {
                Button(
                    ctx.config.strings.openGithubIssues,
                    systemImage: "link",
                    action: openGitHubIssues
                )
                Button(
                    ctx.config.strings.sendEmail,
                    systemImage: "envelope",
                    action: sendBugEmail
                )
            } header: {
                Label(ctx.config.strings.reportBug, systemImage: "ladybug")
            }
        }

        // MARK: - Layer 2: data-driven variant picker

        @ViewBuilder
        private func variantSection(selectedVariantId: Binding<String>) -> some View {
            let variants = ctx.activeProfile.variants
            if variants.count > 1 {
                Section {
                    Picker(
                        ctx.activeProfile.displayName,
                        selection: selectedVariantId
                    ) {
                        ForEach(variants, id: \.id) { variant in
                            Text(variant.displayName).tag(variant.id)
                        }
                    }
                    .pickerStyle(.segmented)
                } header: {
                    Label(ctx.activeProfile.displayName, systemImage: "textformat")
                }
                .onChange(of: ctx.settings.selectedVariantId) { _, _ in
                    ctx.rebuildEngine()
                }
            }
        }

        // MARK: - Layer 3: injectable descriptors

        private func descriptorSections() -> some View {
            ForEach(ctx.config.extraSettings) { descriptor in
                Section {
                    descriptorControl(descriptor)
                } footer: {
                    if let footer = descriptor.footer {
                        Text(footer)
                    }
                }
            }
        }

        @ViewBuilder
        private func descriptorControl(_ descriptor: SettingDescriptor) -> some View {
            switch descriptor.kind {
            case .toggle:
                Toggle(
                    descriptor.label,
                    isOn: toggleBinding(for: descriptor)
                )
            case let .picker(options):
                Picker(
                    descriptor.label,
                    selection: pickerBinding(for: descriptor)
                ) {
                    ForEach(options, id: \.self) { option in
                        Text(option).tag(option)
                    }
                }
            }
        }

        private func toggleBinding(for descriptor: SettingDescriptor) -> Binding<Bool> {
            Binding(
                get: {
                    (ctx.settings.value(for: descriptor.key) ?? descriptor.defaultValue) == "true"
                },
                set: { newValue in
                    ctx.settings.setValue(newValue ? "true" : "false", for: descriptor.key)
                }
            )
        }

        private func pickerBinding(for descriptor: SettingDescriptor) -> Binding<String> {
            Binding(
                get: {
                    ctx.settings.value(for: descriptor.key) ?? descriptor.defaultValue
                },
                set: { newValue in
                    ctx.settings.setValue(newValue, for: descriptor.key)
                }
            )
        }

        // MARK: - DEBUG section

        #if DEBUG
            private var debugSection: some View {
                Section {
                    Button("Reset review prompt", systemImage: "arrow.counterclockwise") {
                        ctx.settings.reviewPromptDismissed = false
                        ctx.settings.reviewLaunchCount = 0
                        showResetToast = true
                    }
                } header: {
                    Label("Debug", systemImage: "hammer")
                } footer: {
                    Text("Clears dismissed flag and launch count. Restart the app to re-trigger after 3 launches.")
                }
            }
        #endif

        // MARK: - Bug-report actions

        private func openGitHubIssues() {
            let repo = ctx.config.branding.githubRepo
            guard let url = URL(string: "https://github.com/\(repo)/issues/new") else {
                return
            }
            openURL(url)
        }

        private func sendBugEmail() {
            let subject = "[iOS] Bug report – \(ctx.config.branding.appName)"
            let body = """
            Description:

            Step to reproduce:

            Device:
            iOS version:
            """
            .replacingOccurrences(of: "\n", with: "\r\n")

            guard
                let encodedSubject = subject
                .addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
                let encodedBody = body
                .addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed),
                let url = URL(
                    string: "mailto:\(ctx.config.branding.supportEmail)"
                        + "?subject=\(encodedSubject)&body=\(encodedBody)"
                )
            else { return }
            openURL(url)
        }
    }
#endif
